PEP 597: Apply grammar, syntax and polish fixes, and clarify phrasing and terminology (#1887)
python/peps: pep-0597.rst

Developers using macOS or Linux may forget that the default encoding
is not always UTF-8.

For example, using ``long_description = open("README.md").read()`` in
``setup.py`` is a common mistake. Many Windows users cannot install

> such packages if there is at least one non-ASCII character (e.g. emoji)

Author: Another common example would be the copyright symbol; not sure if its worth adding

Reviewer: In my personal experience, author/contributor names are more common source of non-ASCII characters.

Author: ```suggestion
such packages if there is at least one non-ASCII character
(e.g. emoji, author names, copyright symbols, and the like)
```

Good point, thanks. What about this?

Reviewer: LGTM

Author: :+1: Feel free to apply

Author: Okay, thanks! I'll apply once you :+1: everything (I forgot this was my PR and not yours, sorry!)

> in their UTF-8-encoded ``README.md`` file.

Author: Here you previously mentioned loading "TOML", but I was very confused what you were referring to as it didn't connect to the previous sentance and the only commonly-used TOML file I'm aware of in packaging is `pyproject.toml`, which AFAIK is not loaded programmatically in setup.py. Is there something I'm missing here?

Of the 4000 most downloaded packages from PyPI, 489 use non-ASCII
characters in their README, and 82 of these do not explicitly specify
the encoding used to open it. [1]_ Therefore, they cannot be installed

> from source in non-UTF-8 locales.

Author: Here you previously mentioned loading "TOML", but I was very confused what you were referring to as it didn't connect to the previous sentance and the only commonly-used TOML file I'm aware of in packaging is `pyproject.toml`, which AFAIK is not loaded programmatically in setup.py. Is there something I'm missing here?

Reviewer: The bug in PEP 517 backend. And it is still alive. See here. https://github.com/takluyver/intreehooks/blob/4d6bdc6445cc84fb9774d0084639e09b308d2615/intreehooks.py#L31-L34

One year ago, `pip install poetry --no-binary :all:` failed in C locale because of this bug. But this backend is not used for now so it works for now.

Author: Ah, I see, thanks. Does the 82 uninstallable packages include those affected by that upstream bug, or is that just counting those with issues in their own `setup.py` (e.g. loading long_description from README)? Is it worth describing in enough detail that readers can understand what you're referring to, or do you have somewhere you can link it?

Reviewer: They are included.

Author: Gotcha, so do we want to mention it with least link it? I couldn't find a canonical link to the upstream issue you mentioned, just downstream reports, so help there would be appreciated. What about revising the last three lines to read like this (I can't make it as a suggestion here, but if you agree I can do it separately):

```rst
characters in their README, and 82 fail to install on non-UTF-8 locales
caused by not specifying a file encoding [1]_, either loading the README
or loading the ``pyproject.toml`` in the PEP 517 build backend [LINK]_.
```

Reviewer: Do we need mentioning all details here?

Intreehooks caused error: poetry
Packages using poetry during install: xdg, flake8-broken-line

pbr.packaging uses the default encoding. It caused install error of: stestr, 
https://opendev.org/openstack/pbr/src/branch/master/pbr/packaging.py#L826-L830

All other errors are caused by reading README or CHANGELOG in setup.py

Concrete logs is here: https://gist.github.com/methane/697ea72ce1cb62f1dbf8da7aaca5a6e9

Author: ```
characters in their README, and 82 fail to install from source on
non-UTF-8 locales due to not specifying an encoding for a non-ASCII
file. [1]_
```

Totally up to you how detailed you want it; I just included that based on what you said previously, sorry. Do you prefer the above? Or something else?

Author: Applied the above as a suggestion, thanks!

> characters in their README, and 82 of these do not explicitly specify
> the encoding used to open it. [1]_ Therefore, they cannot be installed
> from source in non-UTF-8 locales.

Author: ```suggestion
characters in their README, and 82 fail to install from source on
non-UTF-8 locales due to not specifying an encoding for a non-ASCII
file. [1]_
```

Per the [discussion](https://github.com/python/peps/pull/1887/files#r600248579) previously, this should be the last needed change; lmk if you want a different proposed version or it to say something else. Once this is LGTM, I'll batch commit everything in and this should be good to merge, thanks.

Reviewer: lgtm

Another example is ``logging.basicConfig(filename="log.txt")``.
Some users might expect it to use UTF-8 by default, but the locale
encoding is actually what is used. [2]_

Even Python experts may assume that the default encoding is UTF-8.
This creates bugs that only happen on Windows; see [3]_, [4]_, [5]_,
and [6]_ for example.

Emitting a warning when the ``encoding`` argument is omitted will help
find such mistakes.

Explicit way to use locale-specific encoding
--------------------------------------------

``open(filename)`` isn't explicit about which encoding is expected:

> * If ASCII is assumed, this isn't a bug, but is inefficient on Windows

Author: Could you clarify what you mean by "inefficient on Windows"? Per the [docs](https://docs.python.org/3/library/codecs.html#standard-encodings), both Latin-1 (the default for a plurality/majority of Windows machines) and `mbcs` (defined as the locale default) take the fast path. Is there a significant speed penalty Latin-1 vs. ASCII, or what is this referring to?

Reviewer: On my Windows environment, cp932 is used and it is much slower than UTF-8.
Maybe, on Windows environments using latin-1 may be not cause slowdown.

Author: ```suggestion
* If ASCII is assumed, this isn't a bug, but may result in decreased
  performance on Windows, particularly with non-Latin-1 locale encodings
```

Gotcha. What about this?

Reviewer: LGTM

Author: :+1: Feel free to apply it

* If the locale encoding is assumed, the behavior is as expected
  (but could change if future versions of Python modify the default)

From this point of view, ``open(filename)`` is not readable code.

``encoding=locale.getpreferredencoding(False)`` can be used to
specify the locale encoding explicitly, but it is too long and easy
to misuse (e.g. one can forget to pass ``False`` as its argument).

This PEP provides an explicit way to specify the locale encoding.

Prepare to change the default encoding to UTF-8
-----------------------------------------------

Since UTF-8 has become the de-facto standard text encoding,
we might default to it for opening files in the future.

However, such a change will affect many applications and libraries.
If we start emitting ``DeprecationWarning`` everywhere the ``encoding``
argument is omitted, it will be too noisy and painful.

Although this PEP doesn't propose changing the default encoding,
it will help enable that change by:

* Reducing the number of omitted ``encoding`` arguments in libraries
  before we start emitting a ``DeprecationWarning`` by default.

* Allowing users to pass ``encoding="locale"`` to suppress

>   the warning, starting with Python 3.10.

Author: I wasn't 100% sure what you meant here by "dropping Python 3.10 support", especially since its not even released yet. I'm guessing you were referring to a hypothetical future version that didn't default to the locale encoding? Not 100% sure how this would change that...?

Reviewer: > I'm guessing you were referring to a hypothetical future version that didn't default to the locale encoding?

I meant hypothetical future version that emits DeprecationWarning by default, not opt-in for using default encoding.
If we don't add `encoding="locale"` in 3.10, it is difficult to suppress the DeprecationWarning without breaking Python 3.10+ compatibility.

Author: Understood, thanks. What about this?

```suggestion
  the current warning and any `DeprecationWarning` added in the future,
  as well as retaining consistent behavior if later Python versions
  change the default, ensuring support for any Python version >=3.10.
```

Reviewer: LGTM, but use ``` ``DeprecationWarning`` ```

Author: ```suggestion
  the current warning and any ``DeprecationWarning`` added in the future,
  as well as retaining consistent behavior if later Python versions
  change the default, ensuring support for any Python version >=3.10.
```

Oops, good catch, thanks! Funny enough I added some additional checks for that in PR #1886 , but I made the same mistake myself, ha!

Specification
=============

``EncodingWarning``
-------------------

Add a new ``EncodingWarning`` warning class as a subclass of
``Warning``. It is emitted when the ``encoding`` argument is omitted and
the default locale-specific encoding is used.

Options to enable the warning
-----------------------------

The ``-X warn_default_encoding`` option and the
``PYTHONWARNDEFAULTENCODING`` environment variable are added. They
are used to enable ``EncodingWarning``.

> ``sys.flags.encoding_warning`` is also added. The flag is ``True`` when

Author: This is correct, right? The literal value of the flag is `True` (haven't gone over the PR yet)

Reviewer: No. Current pull request uses ``1``.

Author: Okay, thanks. I assume that's to be consistent with C and the other flags? Suggestion updates it accordingly.

```suggestion
``sys.flags.encoding_warning`` is also added. The flag is ``1`` when
```

Reviewer: I intentionally use the term `true` because ``True`` or ``1`` is just implementation details and sys.flags don't guarantee the types in general. We may change the type without any deprecation.

https://docs.python.org/ja/3/library/sys.html#sys.flags

Author: ```suggestion
``sys.flags.encoding_warning`` is also added. The flag is truthy when
```

Understood, thanks. I revised it to the standard term "truthy" to clarify more explicitly that users can only rely on it being truthy (i.e. `== True`) and don't mistakenly think "true" refers to the `True` literal.

Reviewer: "true" is common term in Python doc.
 
```
$ rg -w true | wc -l
     667

$ rg truthy
reference/compound_stmts.rst
588:   * If the guard evaluates as truthy or missing, the ``block`` inside ``case_block`` is
640:   * If the ``guard`` condition evaluates to "truthy", the case block is

library/ast.rst
1263:    is truthy, the body of the case shall be executed. ``body`` contains a list
1264:    of nodes to execute if the guard is truthy.
```

Author: ```suggestion
``sys.flags.encoding_warning`` is also added. The flag is true when
```

Okay, thanks. Updated accordingly.

Author: Applied @methane 's suggestion instead.

> ``sys.flags.encoding_warning`` is also added. The flag is ``True`` when

Reviewer: ```suggestion
``sys.flags.warn_default_encoding`` is also added. The flag is true when
```

``EncodingWarning`` is enabled.

When the flag is set, ``io.TextIOWrapper()``, ``open()`` and other
modules using them will emit ``EncodingWarning`` when the ``encoding``
argument is omitted.

Since ``EncodingWarning`` is a subclass of ``Warning``, they are

> shown by default (if the ``encoding_warning`` flag is set), unlike

Author: ```suggestion
shown by default (if the ``warn_default_encoding`` flag is set), unlike
```

Reviewer: lgtm

``DeprecationWarning``.

``encoding="locale"``
---------------------

``io.TextIOWrapper`` will accept ``"locale"`` as a valid argument to
``encoding``. It has the same meaning as the current ``encoding=None``,
except that ``io.TextIOWrapper`` doesn't emit ``EncodingWarning`` when
``encoding="locale"`` is specified.

``io.text_encoding()``
----------------------

``io.text_encoding()`` is a helper for functions with an
``encoding=None`` parameter that pass it to ``io.TextIOWrapper()`` or
``open()``.

A pure Python implementation will look like this::

   def text_encoding(encoding, stacklevel=1):
       """A helper function to choose the text encoding.

       When *encoding* is not None, just return it.
       Otherwise, return the default text encoding (i.e. "locale").

       This function emits an EncodingWarning if *encoding* is None and

>        sys.flags.encoding_warning is True.

Author: ```suggestion
       sys.flags.warn_default_encoding is true.
```

Reviewer: lgtm

       This function can be used in APIs with an encoding=None parameter
       that pass it to TextIOWrapper or open.
       However, please consider using encoding="utf-8" for new APIs.
       """
       if encoding is None:

>            if sys.flags.encoding_warning:

Author: ```suggestion
           if sys.flags.warn_default_encoding:
```

Reviewer: lgtm

               import warnings
               warnings.warn(

>                    "The 'encoding' argument was omitted; please "
>                    "explicitly specify one (e.g. 'utf-8'), or 'locale' "
>                    "to use the current system's locale encoding",

Author: ```suggestion
                   "The 'encoding' argument was omitted; please "
                   "explicitly specify one (e.g. 'utf-8'), or 'locale' "
                   "to use the current system's locale encoding",
```

If the extra detail here on how to respond to this is not desired/appropriate, everything after the semicolon `;` can be removed

Reviewer: Yes. They are removed already in the pull request: https://github.com/python/cpython/pull/19481

Author: ```suggestion
                   "'encoding' argument not specified.",
```

Thanks, updated to match python/cpython#19481

Reviewer: LGTM
